singer/src/main/java/com/pinterest/singer/writer/s3/S3Writer.java

```diff
@@ -0,0 +1,438 @@
+     *
+     * @param isDraining whether the system is in a draining state
+     * @throws LogStreamWriterException if an error occurs while creating or writing to the buffer file
+     */
+    @Override
+    public synchronized void startCommit(boolean isDraining) throws LogStreamWriterException {
+        messageCount = 0;
+        try {
+            if (!bufferFile.exists()) {
+                bufferFile.createNewFile();
+            }
+
+            bufferedOutputStream = new BufferedOutputStream(new FileOutputStream(bufferFile, true));
+
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to create buffer file: " + bufferFile.getName(), e);
+        }
+        if (uploadFuture == null) {
+            scheduleUploadTask();
+        }
+    }
+
+
+    /**
+     * Schedules a task to upload the buffer file at regular intervals.
+     * If the buffer file exists and has data, it is renamed and a new buffer file is created.
+     * The renamed file is then uploaded to S3 (or a similar storage service).
+     */
+    private void scheduleUploadTask() {
+        synchronized (objLock) {
+            if (uploadFuture != null && !uploadFuture.isDone()) {
+                LOG.info("An upload task is already scheduled or running");
+                return;
+            }
+            uploadFuture = fileUploadTimer.schedule(() -> {
+                try {
+                    synchronized (objLock) {
+                        if (bufferFile.length() > 0) {
+                            File newFile = createNewS3File();
+                            String fileFormat = getS3FileFormat();
+                            bufferedOutputStream.close();
+                            uploadDiskBufferedFileToS3(newFile, fileFormat);
+                        }
+                        uploadFuture = null;
+                    }
+                } catch (Exception e) {
+                    LOG.error("Failed to upload buffer file to S3", e);
+                }
+            }, minUploadTime, TimeUnit.SECONDS);
+        }
+    }
+
+    /**
+     * Helper function that uploads the disk buffered file to s3
+     * */
+    private void uploadDiskBufferedFileToS3(File newFile, String fileFormat) throws IOException {
+        if (bufferFile.renameTo(newFile)) {
+            String bufferFileName = sanitizeFileName(logStream.getFullPathPrefix()) + ".buffer.log";
+            bufferFile = new File(BUFFER_DIR, bufferFileName);
+            bufferFile.createNewFile();
+            bufferedOutputStream = new BufferedOutputStream(new FileOutputStream(bufferFile, true));
+            uploadAndRecordMetrics(newFile, fileFormat);
+            newFile.delete();
+        } else {
+            LOG.error("Failed to rename buffer file");
+        }
+    }
+
+
+    /**
+     * Helper function that uploads the file to s3 and records metrics
+     * */
+    private void uploadAndRecordMetrics(File newFile, String fileFormat) {
```

Comment:
Can we add a `s3` after "writer" in the metric name to signify the writer type? E.g "singer.writer.s3.num_uploads" or "singer.writer.s3.num_failed_uploads"

Reply:
ack, I have changed these